### PR TITLE
[CIT-270] configure moquitto port

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR captures the code changes to add a feature to configure a port for mosquito.  Also to inform the mqtt clients to use the newer port.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/src/SUMMARY.md [hunk 1]
```diff
@@ -16,6 +16,7 @@
     - [How to connect a cloud end-point](./howto-guides/004_connect.md)
     - [How to send MQTT messages](./howto-guides/005_pub_sub.md)
     - [How to test connection to cloud](./howto-guides/007_test_connectiom.md)
+    - [How to configure mqtt port](./howto-guides/008_config_mqtt_port.md)
```
Comment (didier-wenzek) [inline docs/src/SUMMARY.md:19]: The title must say of which port are we speaking.

```
How to configure the port of the local MQTT bus
```

File: docs/src/howto-guides/008_config_mqtt_port.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+## Step 2: Inform about the new port to mqtt clients
+
+Once the mosquitto daemon is configured to use a specific port. This has to be notified 
+to all the mqtt clients that are using mosquitto to publish/subscribe.
+Follow below step to do that
+   
+```shell
+tedge config set mosquitto.port 1024
+```
+This will make sure that all the mqtt clients use the newer port that has been set.
```
Comment (didier-wenzek) [inline docs/src/howto-guides/008_config_mqtt_port.md:26]: This only ensures that the `tedge` clients are using the appropriate MQTT port.

Any other client, as `collectd`  will have to be configured in its own way.

File: docs/src/howto-guides/008_config_mqtt_port.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+## Verify the port configured/set
+
+Use below command to check the port has been set properly or not.
+ 
+```shell
+tedge config get mosquitto.port
+```
+This prints out the port that has been set.
+
+## Update to use default port
```
Comment (didier-wenzek) [inline docs/src/howto-guides/008_config_mqtt_port.md:37]: What is the default port? It might be good to name it.

File: docs/src/howto-guides/README.md [hunk 4]
```diff
@@ -5,3 +5,5 @@
 3. [How to connect a cloud end-point?](./004_connect.md)
 4. [How to use `tedge mqtt` module?](./005_pub_sub.md)
 5. [How to test connection to cloud?](./007_test_connectiom.md)
+6. [How to configure mqtt port?](./008_config_mqtt_port.md)
```
Comment (didier-wenzek) [inline docs/src/howto-guides/README.md:8]: Same thing as for the `SUMMARY.d`

File: tedge_config/src/settings.rs [hunk 5]
```diff
@@ -126,6 +127,20 @@ impl ConfigSetting for AzureRootCertPathSetting {
     type Value = FilePath;
 }
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct MosquittoPortSetting;
```
Comment (didier-wenzek) [inline tedge_config/src/settings.rs:131]: I would prefer a name that is unrelated to Mosquitto as `MQTTBusPortSetting`.

File: tedge_config/src/tedge_config_defaults.rs [hunk 6]
```diff
@@ -47,6 +50,7 @@ impl From<&TEdgeConfigLocation> for TEdgeConfigDefaults {
                 .into(),
             default_azure_root_cert_path: system_cert_path.clone().into(),
             default_c8y_root_cert_path: system_cert_path.into(),
+            default_mosquitto_port: 1883,
```
Comment (didier-wenzek) [inline tedge_config/src/tedge_config_defaults.rs:53]: This default seems to be unused, since one has a `DEFAULT_PORT` in `mqtt/cli.rs`.

File: tedge/src/cli/mqtt/cli.rs [hunk 7]
```diff
@@ -37,7 +38,18 @@ pub enum TEdgeMqttCli {
...
-    fn build_command(self, _context: BuildContext) -> Result<Box<dyn Command>, crate::ConfigError> {
+    fn build_command(self, context: BuildContext) -> Result<Box<dyn Command>, crate::ConfigError> {
+        let port = match context
+            .config_repository
+            .load()?
+            .query(MosquittoPortSetting)
+        {
+            Ok(p) => p,
+            _ => Port(DEFAULT_PORT),
+        };
```
Comment (rina23q) [inline tedge/src/cli/mqtt/cli.rs:49]: You don't need this match. You already set default in tedge_config library. If `mosquitto.port` is not set by user, `mosquitto.port` will use `1883`.
So, this can be simple and remove `DEFAULT_PORT`.
```suggestion
        let port = context
            .config_repository
            .load()?
            .query(MosquittoPortSetting)?;
```

**Review by rina23q — COMMENTED** [review]
Left some comments. This code can be polished.

File: tedge_config/src/tedge_config_defaults.rs (hunk 6 above)
Comment (rina23q) [reply tedge_config/src/tedge_config_defaults.rs:53]: I'd rather prefer to keep this and remove DEFAULT_PORT and match-sentence in `mqtt/cli.rs`.

File: tedge_config/src/settings.rs [hunk 8]
```diff
@@ -126,6 +127,20 @@ impl ConfigSetting for AzureRootCertPathSetting {
...
+
+impl ConfigSetting for MosquittoPortSetting {
+    const KEY: &'static str = "mosquitto.port";
```
Comment (didier-wenzek) [inline tedge_config/src/settings.rs:134]: This property should be `mqtt.port`.

Thin-edge is currently using mosquitto but is not tied to.

File: docs/src/SUMMARY.md (hunk 1 above)
Comment (PradeepKiruvale) [reply docs/src/SUMMARY.md:19]: done

File: docs/src/howto-guides/008_config_mqtt_port.md (hunk 3 above)
Comment (PradeepKiruvale) [reply docs/src/howto-guides/008_config_mqtt_port.md:37]: You mean the default port number? 1883?

Comment (didier-wenzek) [reply docs/src/howto-guides/008_config_mqtt_port.md:37]: yes !

File: common/mqtt_client/src/lib.rs [hunk 9]
```diff
@@ -374,12 +374,12 @@ pub struct Config {
     /// frontend operations.
     ///
     /// Default: `10`.
-    queue_capacity: usize,
+    pub queue_capacity: usize,
```
Comment (didier-wenzek) [inline common/mqtt_client/src/lib.rs:377]: Why do you need to change this visibility?

File: docs/src/SUMMARY.md [hunk 10]
```diff
@@ -16,6 +16,7 @@
     - [How to connect a cloud end-point](./howto-guides/004_connect.md)
     - [How to send MQTT messages](./howto-guides/005_pub_sub.md)
     - [How to test connection to cloud](./howto-guides/007_test_connectiom.md)
+    - [How to configure local mqtt port](./howto-guides/008_config_mqtt_port.md)
```
Comment (didier-wenzek) [inline docs/src/SUMMARY.md:19]: This sounds like a missing word. Why not "How to configure __the__ local mqtt port"?

File: mapper/dm_agent/src/monitor.rs [hunk 11]
```diff
@@ -21,12 +21,12 @@ use std::time::Duration;
...
-    mqtt_client_id: &'static str,
-    mqtt_source_topic: &'static str,
-    mqtt_target_topic: &'static str,
-    batching_window: u64,
+    pub host: &'static str,
+    pub port: u16,
+    pub mqtt_client_id: &'static str,
+    pub mqtt_source_topic: &'static str,
+    pub mqtt_target_topic: &'static str,
+    pub batching_window: u64,
```
Comment (didier-wenzek) [inline mapper/dm_agent/src/monitor.rs:29]: This seems unrelated to that PR

File: common/mqtt_client/src/lib.rs (hunk 9 above)
Comment (PradeepKiruvale) [reply common/mqtt_client/src/lib.rs:377]: When I try to update the port number in the config, the compiler throws an error saying its private, so cant change.

File: mapper/dm_agent/src/monitor.rs (hunk 11 above)
Comment (PradeepKiruvale) [reply mapper/dm_agent/src/monitor.rs:29]: Same as above, while I create the config with the port number, it throws an error telling these struct members are private and cant update.

**Review by mneumann — APPROVED** [review]
Looks top to me. There are some minor improvements that you could do (see comments).

File: tedge_config/src/models/port.rs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    fn try_into(self) -> Result<String, Self::Error> {
+        Ok(format!("{}", self.0))
+    }
+}
+
+impl Into<u16> for Port {
+    fn into(self) -> u16 {
+        self.0
+    }
+}
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply tedge_config/src/models/port.rs:37]: added

File: common/mqtt_client/src/lib.rs (hunk 9 above)
Comment (PradeepKiruvale) [reply common/mqtt_client/src/lib.rs:377]: Added a setter function and reverted it back to private

File: mapper/dm_agent/src/monitor.rs (hunk 11 above)
Comment (PradeepKiruvale) [reply mapper/dm_agent/src/monitor.rs:29]: Implemented a setter function and reverted them back to private

File: tedge/src/cli/mqtt/cli.rs [hunk 13]
```diff
@@ -37,7 +37,11 @@ pub enum TEdgeMqttCli {
 }
 
 impl BuildCommand for TEdgeMqttCli {
-    fn build_command(self, _context: BuildContext) -> Result<Box<dyn Command>, crate::ConfigError> {
+    fn build_command(self, context: BuildContext) -> Result<Box<dyn Command>, crate::ConfigError> {
+        let port = context.config_repository.load()?.query(MqttPortSetting)?;
+
+        let mqtt_config = mqtt_client::Config::new(DEFAULT_HOST, port.into());
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply tedge/src/cli/mqtt/cli.rs:43]: Not part of this PR, or at least not needed as of now because its always localhost.

File: tedge_config/src/models/port.rs [hunk 14]
```diff
@@ -0,0 +1,37 @@
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Debug, Copy, Clone, serde::Serialize, serde::Deserialize, Eq, PartialEq)]
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply tedge_config/src/models/port.rs:3]: ok, I removed them

**Review by rina23q — APPROVED** [review]

File: mapper/dm_agent/src/main.rs [hunk 15]
```diff
@@ -33,3 +39,33 @@ async fn main() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+fn get_mqtt_port() -> Result<u16, anyhow::Error> {
+    let config_repository = get_config_repository()?;
+    let tedge_config = config_repository.load()?;
+    Ok(tedge_config.query(MqttPortSetting)?.0)
```
[the remark this answers is not in the document]
Comment (mneumann) [reply mapper/dm_agent/src/main.rs:44]: NEW: Use .into() instead .0 here

File: mapper/tedge_mapper/src/main.rs [hunk 16]
```diff
@@ -91,6 +84,15 @@ async fn main() -> anyhow::Result<()> {
     Ok(())
 }
 
+fn get_mqtt_config() -> Result<Config, anyhow::Error> {
```
Comment (mneumann) [inline mapper/tedge_mapper/src/main.rs:87]: NEW: please also remove the `get_` prefix here.

File: mapper/tedge_mapper/src/main.rs [hunk 17]
```diff
@@ -91,6 +84,15 @@ async fn main() -> anyhow::Result<()> {
...
+    Ok(Config::default().with_port(get_tedge_config()?.query(MqttPortSetting)?.into()))
+}
+
+fn get_tedge_config() -> Result<TEdgeConfig, anyhow::Error> {
```
Comment (mneumann) [inline mapper/tedge_mapper/src/main.rs:91]: NEW: here as well

File: tedge_config/src/models/port.rs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+impl Into<u16> for Port {
+    fn into(self) -> u16 {
+        self.0
+    }
+}
+
+#[cfg(test)]
+use assert_matches::*;
+#[test]
+
```
Comment (mneumann) [inline tedge_config/src/models/port.rs:41]: NEW: please remove empty line. add empty line before `#[test]`.

**Review by mneumann — COMMENTED** [review]
Some minor changes See "NEW". Also, please address some of Didier's suggestions.

**Review by mneumann — APPROVED** [review]
